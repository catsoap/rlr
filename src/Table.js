import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { sortBy } from 'lodash';
import { Button } from './Button';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

class Table extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey) {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  }

  render() {
    const { list, onDismiss } = this.props;
    const { sortKey, isSortReverse } = this.state;
    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="pa-2 font-semibold">
        <div>
          <span className="pa-1">
            <Sort sortKey={'TITLE'} onSort={this.onSort} activeSortKey={sortKey}>
              Title
            </Sort>
          </span>
          <span className="pa-1">
            <Sort sortKey={'AUTHOR'} onSort={this.onSort} activeSortKey={sortKey}>
              Author
            </Sort>
          </span>
          <span className="pa-1">
            <Sort sortKey={'COMMENTS'} onSort={this.onSort} activeSortKey={sortKey}>
              Comments
            </Sort>
          </span>
          <span className="pa-1">
            <Sort sortKey={'POINTS'} onSort={this.onSort} activeSortKey={sortKey}>
              Points
            </Sort>
          </span>
          <span className="pa-1">Archive</span>
        </div>
        <div>
          {reverseSortedList.map(item => (
            <div key={item.objectID}>
              <span className="pa-1">
                <a href={item.url}>{item.title}</a>
              </span>
              <span className="pa-1">{item.author}</span>
              <span className="pa-1">{item.num_comments}</span>
              <span className="pa-1">{item.points}</span>
              <span className="pa-1">
                <Button
                  onClick={() => onDismiss(item.objectID)}
                  className="pa-1 rounded font-semibold text-gray-700"
                >
                  Dismiss
                </Button>
              </span>
            </div>
          ))}
        </div>
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    }),
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
};

const Sort = ({ sortKey, activeSortKey, onSort, children }) => {
  const sortClass = classNames('pa-1 rounded font-semibold text-gray-700', {
    red: sortKey === activeSortKey,
  });

  return (
    <Button onClick={() => onSort(sortKey)} className={sortClass}>
      {children}
    </Button>
  );
};

export default Table;
