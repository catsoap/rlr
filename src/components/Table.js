/** @jsx jsx */
// eslint-disable-next-line
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import formatDistance from 'date-fns/formatDistance';
import { sortBy } from 'lodash';
import { ReactComponent as DismissIcon } from '../svg/dismiss.svg';
import { ReactComponent as UrlIcon } from '../svg/url.svg';
// eslint-disable-next-line
import { jsx, Button, IconButton, Link } from 'theme-ui';

const SORTS = {
  NONE: (list) => list,
  TITLE: (list) => sortBy(list, 'title'),
  AUTHOR: (list) => sortBy(list, 'author'),
  COMMENTS: (list) => sortBy(list, 'num_comments').reverse(),
  POINTS: (list) => sortBy(list, 'points').reverse(),
  CREATEDAT: (list) => sortBy(list, 'created_at_i'),
};

class Table extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey) {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  }

  render() {
    const { list, onDismiss } = this.props;
    const { sortKey, isSortReverse } = this.state;
    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;
    const now = new Date();

    return (
      <section>
        <header sx={{ display: ['none', 'flex'], mb: 3 }}>
          <div sx={{ flex: 16 }}>
            <Sort sortKey={'TITLE'} onSort={this.onSort} activeSortKey={sortKey}>
              Title
            </Sort>
          </div>
          <div sx={{ flex: 3 }}>
            <Sort sortKey={'AUTHOR'} onSort={this.onSort} activeSortKey={sortKey}>
              Author
            </Sort>
          </div>
          <div sx={{ flex: 1 }}>
            <Sort sortKey={'COMMENTS'} onSort={this.onSort} activeSortKey={sortKey}>
              Coms
            </Sort>
          </div>
          <div sx={{ flex: 1 }}>
            <Sort sortKey={'POINTS'} onSort={this.onSort} activeSortKey={sortKey}>
              Pts
            </Sort>
          </div>
          <div sx={{ flex: 2 }}>
            <Sort sortKey={'CREATEDAT'} onSort={this.onSort} activeSortKey={sortKey}>
              Date
            </Sort>
          </div>
          <div sx={{ flex: 1 }}></div>
        </header>

        {reverseSortedList.map((item, index) => (
          <div key={index} sx={{ display: 'flex', fontSize: '.85em' }}>
            <div sx={{ flex: 16 }}>
              <Link
                href={`https://news.ycombinator.com/item?id=${item.objectID}`}
                title={item.title ? 'Read comments' : 'Read comment'}
                target="_blank"
                rel="noopener noreferrer"
              >
                {item.title ? item.title : item.comment_text.substring(0, 70) + '...'}
              </Link>
              {item.url && (
                <IconButton
                  onClick={() => window.open(item.url, '_blank')}
                  title={`Go to original post: ${item.url}`}
                >
                  <UrlIcon
                    sx={{
                      fill: 'cornflowerblue',
                      width: '1.25em',
                      height: '1.25em',
                    }}
                  />
                </IconButton>
              )}
            </div>
            <div sx={{ flex: 3, display: ['none', 'block'] }}>{item.author}</div>
            <div sx={{ flex: 1, display: ['none', 'block'] }}>{item.num_comments} </div>
            <div sx={{ flex: 1, display: ['none', 'block'] }}>{item.points}</div>
            <div sx={{ flex: 2, display: ['none', 'block'] }}>
              {formatDistance(new Date(item.created_at_i * 1000), now, { addSuffix: true })}
            </div>
            <div sx={{ flex: 1, display: ['none', 'block'] }}>
              <IconButton onClick={() => onDismiss(item.objectID)}>
                <DismissIcon
                  sx={{
                    fill: 'danger',
                    width: '1em',
                    height: '1em',
                  }}
                />
              </IconButton>
            </div>
          </div>
        ))}
      </section>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    }),
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
};

const Sort = ({ sortKey, activeSortKey, onSort, children }) => {
  const colors =
    sortKey === activeSortKey
      ? {
          bg: 'muted',
          borderColor: 'primary',
        }
      : {
          bg: 'background',
        };

  return (
    <Button onClick={() => onSort(sortKey)} sx={{ variant: 'buttons.sorter', ...colors }}>
      {children}
    </Button>
  );
};

export default Table;
