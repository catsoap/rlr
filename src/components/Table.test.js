import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Table from './Table';

Enzyme.configure({ adapter: new Adapter() });

describe('Table', () => {
  const props = {
    list: [
      {
        title: '1',
        author: '1',
        num_comments: 1,
        points: 2,
        objectID: 'y',
        created_at_i: 1453762117,
      },
      {
        title: '2',
        author: '2',
        num_comments: 1,
        points: 2,
        objectID: 'z',
        created_at_i: 1453762117,
      },
    ],
    onDismiss: () => 'ok',
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in list', () => {
    const element = shallow(<Table {...props} />);
    expect(element.find('* + section > div').length).toBe(2);
  });
});
