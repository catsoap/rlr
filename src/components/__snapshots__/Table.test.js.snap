// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot 1`] = `
<section>
  <header
    className="css-17n9c24"
  >
    <div
      className="css-mu2wql"
    >
      <button
        className="css-12rur1d-Sort"
        onClick={[Function]}
      >
        Title
      </button>
    </div>
    <div
      className="css-ieri5n"
    >
      <button
        className="css-12rur1d-Sort"
        onClick={[Function]}
      >
        Author
      </button>
    </div>
    <div
      className="css-1rr4qq7"
    >
      <button
        className="css-12rur1d-Sort"
        onClick={[Function]}
      >
        Coms
      </button>
    </div>
    <div
      className="css-1rr4qq7"
    >
      <button
        className="css-12rur1d-Sort"
        onClick={[Function]}
      >
        Pts
      </button>
    </div>
    <div
      className="css-kzfr2u"
    >
      <button
        className="css-12rur1d-Sort"
        onClick={[Function]}
      >
        Date
      </button>
    </div>
    <div
      className="css-1rr4qq7"
    />
  </header>
  <div
    className="css-atfqj8"
  >
    <div
      className="css-mu2wql"
    >
      <a
        className="css-vurnku"
        href="https://news.ycombinator.com/item?id=y"
        rel="noopener noreferrer"
        target="_blank"
        title="Read comments"
      >
        1
      </a>
    </div>
    <div
      className="css-q74yqm"
    >
      1
    </div>
    <div
      className="css-14ybhvo"
    >
      1
       
    </div>
    <div
      className="css-14ybhvo"
    >
      2
    </div>
    <div
      className="css-wokw6q"
    >
      about 4 years ago
    </div>
    <div
      className="css-14ybhvo"
    >
      <button
        className="css-wwcqj1"
        onClick={[Function]}
      >
        <svg
          className="css-1g5ihon"
        >
          dismiss.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    className="css-atfqj8"
  >
    <div
      className="css-mu2wql"
    >
      <a
        className="css-vurnku"
        href="https://news.ycombinator.com/item?id=z"
        rel="noopener noreferrer"
        target="_blank"
        title="Read comments"
      >
        2
      </a>
    </div>
    <div
      className="css-q74yqm"
    >
      2
    </div>
    <div
      className="css-14ybhvo"
    >
      1
       
    </div>
    <div
      className="css-14ybhvo"
    >
      2
    </div>
    <div
      className="css-wokw6q"
    >
      about 4 years ago
    </div>
    <div
      className="css-14ybhvo"
    >
      <button
        className="css-wwcqj1"
        onClick={[Function]}
      >
        <svg
          className="css-1g5ihon"
        >
          dismiss.svg
        </svg>
      </button>
    </div>
  </div>
</section>
`;
